In [17]:
import pandas as pd


**Loading WW2 Ships Dataset**

In [29]:
df= pd.read_csv("ships.csv")
df

,Unnamed: 0,Name,Class,Country,Launch Year
0,0,A (Planned),No Classification,Germany,
1,1,ABSD-1,No Classification,United States,
2,2,ABSD-2,No Classification,United States,
3,3,Abukuma,Nagara-class Light Cruiser,Japan,1923
4,4,Acasta,A and B-class Destroyer,United Kingdom,1929
...,...,...,...,...,...
1781,1781,Zara,Zara-class Heavy Cruiser,Italy,1930
1782,1782,Zhaohe,Zhaohe-class Protected Cruiser,China,1911
1783,1783,Zuiho,Zuiho-class Light Carrier,Japan,1936
1784,1784,Zuikaku,Shokaku-class Aircraft Carrier,Japan,1939


**Checking the features of the loaded dataset**

In [26]:
print('''
****Here is the description of Dataset****
''')
print(df.describe())
print('''
****Here is the info of Dataset****
''')
df.info()
print('''
****Here is the null value sum up of Dataset****
''')
print(df.isnull().sum())
print('''
****Here is the shape and column names of Dataset****
''')
df.shape
df.columns


****Here is the description of Dataset****

        Unnamed: 0
count  1786.000000
mean    892.500000
std     515.718108
min       0.000000
25%     446.250000
50%     892.500000
75%    1338.750000
max    1785.000000

****Here is the info of Dataset****

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1786 entries, 0 to 1785
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Unnamed: 0   1786 non-null   int64 
 1   Name         1786 non-null   object
 2   Class        1786 non-null   object
 3   Country      1786 non-null   object
 4   Launch Year  1786 non-null   object
dtypes: int64(1), object(4)
memory usage: 69.9+ KB

****Here is the null value sum up of Dataset****

Unnamed: 0     0
Name           0
Class          0
Country        0
Launch Year    0
dtype: int64

****Here is the shape and column names of Dataset****



Index(['Unnamed: 0', 'Name', 'Class', 'Country', 'Launch Year'], dtype='object')

In [39]:
df.Country.head(10)
df['Country'].unique()

array(['Germany\xa0', 'United States\xa0', 'Japan\xa0',
       'United Kingdom\xa0', 'Australia\xa0', 'Italy\xa0', 'France\xa0',
       'Canada\xa0', 'Poland\xa0', 'Greece\xa0', 'China\xa0',
       'Finland\xa0', 'Netherlands\xa0', 'Russia\xa0', 'Norway\xa0'],
      dtype=object)

In [59]:
# removing the unnamed index column
df = pd.read_csv('ships.csv')
df = df.drop(columns=[df.columns[0]])
# Cleaning up Whitespace & Special Characters
df = df.apply(lambda col: col.str.strip() if col.dtype == "object" else col)
df['Name'] = df['Name'].str.replace('Lüderitz', 'Lüderitz', regex=False)

# Cleaning 'Launch Year' (Convert empty strings/spaces to actual NaN, then convert to numeric/nullable integer)
df['Launch Year'] = pd.to_numeric(df['Launch Year'].replace('', pd.NA), errors='coerce').astype('Int64')

# Handling Missing Values in Categorical Columns (Replacing empty class/country with 'Unknown')
df['Class'] = df['Class'].fillna('Unknown')
df['Country'] = df['Country'].fillna('Unknown')

# Removing all Duplicates
df = df.drop_duplicates()
# df.head(10)
df.tail(10)

,Name,Class,Country,Launch Year
1775,Z57 (Planned),1944-class Destroyer,Germany,<NA>
1776,Z58 (Planned),1944-class Destroyer,Germany,<NA>
1777,Z6 Theodor Riedel,1934A-class Destroyer,Germany,1936
1778,Z7 Hermann Schoemann,1934A-class Destroyer,Germany,1936
1779,Z8 Bruno Heinemann,1934A-class Destroyer,Germany,1936
1780,Z9 Wolfgang Zenker,1934A-class Destroyer,Germany,1936
1781,Zara,Zara-class Heavy Cruiser,Italy,1930
1782,Zhaohe,Zhaohe-class Protected Cruiser,China,1911
1783,Zuiho,Zuiho-class Light Carrier,Japan,1936
1784,Zuikaku,Shokaku-class Aircraft Carrier,Japan,1939


**Now i have the cleaned version of this dataset**

**Following are the 8 specific business questions about this whole procedure with code sections of each question separately**

1. Which country built the most ships?

In [60]:
df.Country.value_counts()

,count
Country,
Germany,949
United States,473
Japan,154
United Kingdom,81
Italy,41
France,28
Russia,15
Australia,12
China,10


**Answer:** As per the results Germany made the most ships

2. What are the most common ship classes overall?

In [61]:
df.Class.value_counts()

,count
Class,
M-class Minesweeper,293
Type VII-class Submarine,143
Balao-class Submarine,103
Type IX-class Submarine,90
Gato-class Submarine,75
...,...
Type B-class Midget Submarine,1
Type C-class Midget Submarine,1
Type D-class Midget Submarine,1


**Answer:** As per the results most common ship classes overall if we give the top 3 rankings are as follows:
1. M-class Minesweeper
2. Type VII-class Submarine
3. Balao-class Submarine

3. Did shipbuilding activity increase or decrease as the war went on?

In [63]:
df.groupby('Country')['Launch Year'].value_counts()


Country        Launch Year
Australia      1918           3
               1917           2
               1934           2
               1927           1
               1942           1
                             ..
United States  1914           1
               1924           1
               1926           1
               1931           1
               1945           1
Name: count, Length: 173, dtype: Int64

In [74]:
ship_production_by_year = df.groupby('Launch Year').size().reset_index(name='Number of Ships')
ship_production_by_year = ship_production_by_year.sort_values(by='Launch Year')

print(ship_production_by_year)

    Launch Year  Number of Ships
0             0                1
1          1893                1
2          1897                2
3          1898                1
4          1900                1
5          1902                1
6          1905                2
7          1906                2
8          1911                7
9          1912                3
10         1913                9
11         1914                4
12         1915                5
13         1916                4
14         1917                7
15         1918               18
16         1919               14
17         1920               10
18         1921                9
19         1922                9
20         1923               10
21         1924                3
22         1925               13
23         1926               15
24         1927               13
25         1928               11
26         1929               15
27         1930               12
28         1931                6
29        

**Answer:** As per the results the shipbulding activity increased as the war went on, bu in 1945 and 1946 only 4 ships in total were built

4. Which country was building the widest variety of ship classes?

In [67]:
df.groupby('Country')['Class'].value_counts()

Country        Class                        
Australia      Leander-class Light Cruiser      3
               V and W-class Destroyer          3
               County-class Heavy Cruiser       2
               Bathurst-class Corvette          1
               Scott-class Destroyer            1
                                               ..
United States  PT-class Motor Torpedo Boat      1
               Ranger-class Aircraft Carrier    1
               Treasury-class USCG Cutter       1
               Victory-class Merchant Vessel    1
               Wasp-class Aircraft Carrier      1
Name: count, Length: 306, dtype: int64

In [73]:
variety_by_country = df.groupby('Country')['Class'].nunique().sort_values(ascending=False)
print(variety_by_country)

Country
United States     76
Japan             73
Germany           58
United Kingdom    44
France            13
Italy             12
Australia          7
China              6
Russia             5
Poland             4
Greece             3
Canada             2
Finland            1
Norway             1
Netherlands        1
Name: Class, dtype: int64


**Answer:** The country with the highest number of widest ship varieties is USA.

5. What's the most common ship class for each of the top 3 countries?

In [95]:
top_3_countries = variety_by_country.head(3).index.tolist()

print(top_3_countries)

# excluding 'No Classification'
for country in top_3_countries:
    country_df = df[df['Country'] == country]

    filtered_classes = country_df['Class'][country_df['Class'] != 'No Classification']

    if not filtered_classes.empty:
        most_common_class = filtered_classes.value_counts().idxmax()
        count_of_class = filtered_classes.value_counts().max()
        print(f"The most common ship class for {country} is '{most_common_class}' with {count_of_class} units.\n")
    else:
        print(f"No other classified ships found for {country} besides 'No Classification'.\n")

['United States', 'Japan', 'Germany']
The most common ship class for United States is 'Balao-class Submarine' with 103 units.

The most common ship class for Japan is 'Nagara-class Light Cruiser' with 6 units.

The most common ship class for Germany is 'M-class Minesweeper' with 293 units.



**Answer:** The most common ship class for each of the top 3 countries (United States, Japan, and Germany) is displayed above.

**Answer:**

6. How much missing data is there in Launch Year, and does it differ by country?

In [91]:
total_missing_launch_year = df['Launch Year'].isnull().sum()
print(total_missing_launch_year)
missing_launch_year_by_country = df.groupby('Country')['Launch Year'].apply(lambda x: x.isnull().sum()).sort_values(ascending=False)
display(missing_launch_year_by_country)

717


,Launch Year
Country,
Germany,472
United States,168
Japan,31
Italy,18
United Kingdom,17
France,4
China,3
Australia,2
Greece,2


**Answer:** The total number of missing 'Launch Year' values is **717** and their distribution by country are displayed above. This shows how much missing data there is and **Germany** contributes the most to it as if we talk about the most missing values of launch year countrywise.

7. Which country had the earliest and latest ship launches on record?

In [106]:
print(df['Launch Year'])
min_yr= df['Launch Year'].min()
print(min_yr)
max_yr= df['Launch Year'].max()
print(max_yr)

valid_launch_years = df['Launch Year'].dropna()
valid_launch_years = valid_launch_years[valid_launch_years != 0]

min_yr = valid_launch_years.min()
max_yr = valid_launch_years.max()
print('''
Min_yr & Max_yr after dropping null values:
''')
print(min_yr)
print(max_yr)

earliest_launch_country = df[df['Launch Year'] == min_yr]['Country'].iloc[0]
latest_launch_country = df[df['Launch Year'] == max_yr]['Country'].iloc[0]

print(earliest_launch_country,min_yr)
print(latest_launch_country,max_yr)

0       <NA>
1       <NA>
2       <NA>
3       1923
4       1929
        ... 
1780    1936
1781    1930
1782    1911
1783    1936
1784    1939
Name: Launch Year, Length: 1778, dtype: Int64
0
1946

Min_yr & Max_yr after dropping null values:

1893
1946
United States 1893
France 1946


**Answer:** USA was the earliest & France is the Latest according to this dataset

8. Which country had the widest range of Launch Years — meaning the longest shipbuilding history in this dataset?

In [112]:
df.describe()

,Launch Year
count,1061.0
mean,1934.381715
std,60.044688
min,0.0
25%,1935.0
50%,1939.0
75%,1942.0
max,1946.0


In [111]:
widest_launch_yrs= df.groupby('Country')['Launch Year']
display(widest_launch_yrs)


In [114]:

cleaned_df = df[df['Launch Year'].notna() & (df['Launch Year'] != 0)].copy()

launch_year_summary = cleaned_df.groupby('Country')['Launch Year'].agg(['min', 'max'])

# Calculating the range
launch_year_summary['range'] = launch_year_summary['max'] - launch_year_summary['min']

sorted_ranges = launch_year_summary.sort_values(by='range', ascending=False)
display(sorted_ranges)


longest_history_country = sorted_ranges.index[0]
longest_history_range = sorted_ranges['range'].iloc[0]


,min,max,range
Country,,,
United States,1893,1945,52
Japan,1898,1944,46
Germany,1905,1945,40
China,1897,1935,38
Russia,1911,1944,33
France,1913,1946,33
United Kingdom,1911,1943,32
Italy,1911,1940,29
Australia,1917,1944,27


**Answer:** The USA had the widest range of Launch Years, indicating the longest shipbuilding history in this dataset.